// nem2-library/dist/buffers/TransactionBuffer.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }(); /*
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      * Copyright 2019 NEM
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      * Licensed under the Apache License, Version 2.0 (the "License");
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      * you may not use this file except in compliance with the License.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      * You may obtain a copy of the License at
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      *     http://www.apache.org/licenses/LICENSE-2.0
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      * Unless required by applicable law or agreed to in writing, software
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      * distributed under the License is distributed on an "AS IS" BASIS,
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      * See the License for the specific language governing permissions and
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      * limitations under the License.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      */

var _BufferUtils = require('./BufferUtils');

var _BufferUtils2 = _interopRequireDefault(_BufferUtils);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var concat_typedarrays = _BufferUtils2.default.concat_typedarrays;
var fit_bytearray = _BufferUtils2.default.fit_bytearray;
var buffer_to_uint = _BufferUtils2.default.buffer_to_uint;
var uint_to_buffer = _BufferUtils2.default.uint_to_buffer;

var TransactionBuffer = function () {
    function TransactionBuffer() {
        var _this = this;

        _classCallCheck(this, TransactionBuffer);

        this.getSize = function () {
            return _this.size;
        };

        this.setSize = function (size) {
            _this.size = size;
        };

        this.getSignature = function () {
            return _this.signature;
        };

        this.setSignature = function (signature) {
            _this.signature = signature;
        };

        this.getSigner = function () {
            return _this.signer;
        };

        this.setSigner = function (signer) {
            _this.signer = signer;
        };

        this.getVersion = function () {
            return _this.version;
        };

        this.setVersion = function (version) {
            _this.version = version;
        };

        this.getType = function () {
            return _this.type;
        };

        this.setType = function (type) {
            _this.type = type;
        };

        this.getFee = function () {
            return _this.fee;
        };

        this.setFee = function (fee) {
            _this.fee = fee;
        };

        this.getDeadline = function () {
            return _this.deadline;
        };

        this.setDeadline = function (deadline) {
            _this.deadline = deadline;
        };

        this.serialize = function () {
            var newArray = new Uint8Array();
            var fitArraysize = fit_bytearray(_this.size, 4);
            newArray = concat_typedarrays(newArray, fitArraysize);
            var fitArraysignature = fit_bytearray(_this.signature, 64);
            newArray = concat_typedarrays(newArray, fitArraysignature);
            var fitArraysigner = fit_bytearray(_this.signer, 32);
            newArray = concat_typedarrays(newArray, fitArraysigner);
            var fitArrayversion = fit_bytearray(_this.version, 2);
            newArray = concat_typedarrays(newArray, fitArrayversion);
            var fitArraytype = fit_bytearray(_this.type, 2);
            newArray = concat_typedarrays(newArray, fitArraytype);
            var fitArrayfee = fit_bytearray(_this.fee, 8);
            newArray = concat_typedarrays(newArray, fitArrayfee);
            var fitArraydeadline = fit_bytearray(_this.deadline, 8);
            newArray = concat_typedarrays(newArray, fitArraydeadline);
            return newArray;
        };
    }

    _createClass(TransactionBuffer, null, [{
        key: 'loadFromBinary',
        value: function loadFromBinary(consumableBuffer) {
            var object = new TransactionBuffer();
            var size = consumableBuffer.get_bytes(4);
            object.size = size;
            var signature = consumableBuffer.get_bytes(64);
            object.signature = signature;
            var signer = consumableBuffer.get_bytes(32);
            object.signer = signer;
            var version = consumableBuffer.get_bytes(2);
            object.version = version;
            var type = consumableBuffer.get_bytes(2);
            object.type = type;
            var fee = consumableBuffer.get_bytes(8);
            object.fee = fee;
            var deadline = consumableBuffer.get_bytes(8);
            object.deadline = deadline;
            return object;
        }
    }]);

    return TransactionBuffer;
}();

var EmbeddedTransactionBuffer = function () {
    function EmbeddedTransactionBuffer() {
        var _this2 = this;

        _classCallCheck(this, EmbeddedTransactionBuffer);

        this.getSize = function () {
            return _this2.size;
        };

        this.setSize = function (size) {
            _this2.size = size;
        };

        this.getSigner = function () {
            return _this2.signer;
        };

        this.setSigner = function (signer) {
            _this2.signer = signer;
        };

        this.getVersion = function () {
            return _this2.version;
        };

        this.setVersion = function (version) {
            _this2.version = version;
        };

        this.getType = function () {
            return _this2.type;
        };

        this.setType = function (type) {
            _this2.type = type;
        };

        this.serialize = function () {
            var newArray = new Uint8Array();
            var fitArraysize = fit_bytearray(_this2.size, 4);
            newArray = concat_typedarrays(newArray, fitArraysize);
            var fitArraysigner = fit_bytearray(_this2.signer, 32);
            newArray = concat_typedarrays(newArray, fitArraysigner);
            var fitArrayversion = fit_bytearray(_this2.version, 2);
            newArray = concat_typedarrays(newArray, fitArrayversion);
            var fitArraytype = fit_bytearray(_this2.type, 2);
            newArray = concat_typedarrays(newArray, fitArraytype);
            return newArray;
        };
    }

    _createClass(EmbeddedTransactionBuffer, null, [{
        key: 'loadFromBinary',
        value: function loadFromBinary(consumableBuffer) {
            var object = new EmbeddedTransactionBuffer();
            var size = consumableBuffer.get_bytes(4);
            object.size = size;
            var signer = consumableBuffer.get_bytes(32);
            object.signer = signer;
            var version = consumableBuffer.get_bytes(2);
            object.version = version;
            var type = consumableBuffer.get_bytes(2);
            object.type = type;
            return object;
        }
    }]);

    return EmbeddedTransactionBuffer;
}();

module.exports = {
    main: TransactionBuffer,
    embedded: EmbeddedTransactionBuffer
};
