// nem2-library/dist/transactions/BaseBuilder.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var BaseBuilder = function () {
    function BaseBuilder() {
        _classCallCheck(this, BaseBuilder);
    }

    _createClass(BaseBuilder, [{
        key: "addFee",
        value: function addFee(fee) {
            this.fee = fee;
            return this;
        }
    }, {
        key: "addVersion",
        value: function addVersion(version) {
            this.version = version;
            return this;
        }
    }, {
        key: "addType",
        value: function addType(type) {
            this.type = type;
            return this;
        }
    }, {
        key: "addDeadline",
        value: function addDeadline(deadline) {
            this.deadline = deadline;
            return this;
        }
    }]);

    return BaseBuilder;
}();

exports.default = BaseBuilder;
